### PR TITLE
AIX: Changes in configuration files to support OpenXL17

### PR DESCRIPTION
**Pull request by suchismith1993**
[Made](url) changes in configuration files to build JDK succesfully for AIX using OpenXLC compiler 17.1.1

### CODE REVIEW

**Comment by karianna** [comment]
@suchismith1993 Are you able to sign the Eclipse CLA?

**Comment by suchismith1993** [comment]
> @suchismith1993 Are you able to sign the Eclipse CLA?

Where can i do that ? It is my first PR for adoptium.

**Comment by smlambert** [comment]
See some ECA guidance here: https://adoptium.net/docs/eca-sign-off

**Comment by suchismith1993** [comment]
> See some ECA guidance here: https://adoptium.net/docs/eca-sign-off

Done.

**Review by karianna — APPROVED** [review]
Code changes look OK, but looking to others to understand if this is truly a new baseline we can work from for all of the users of our AIX binaries.

**Comment by Haroon-Khel** [comment]
Fixes https://github.com/adoptium/aqa-tests/issues/5169

**Comment by sxa** [comment]
> Code changes look OK, but looking to others to understand if this is truly a new baseline we can work from for all of the users of our AIX binaries.

This should be for JDK21+ only

**Review by sxa — CHANGES_REQUESTED** [review]
This will need an update as it seems to revert the changes in https://github.com/adoptium/temurin-build/pull/3809

Also, why has the `-qnotimestamps` parameter been removed - has it been removed from XLC17, and is there an alternative (or is that functionality the default now?)

Also in relation to @karianna's comment  I thought this change was only planned to be done for JDK21+, but it looks like this PR is bumping it up for JDK11+ which I wasn't expecting and would likely have implications on which AIX version we could support.

We should also remove the `set -x` before this is merged.

It would also be good to ensure that JDK8 still builds using the changes in this PR.

**Comment by judovana** [comment]
> This will need an update as it seems to revert the changes in #3809

This is dcone by the automatic merges aka `Merge branch 'master' into openxlcchange`. I hit it recently:  https://github.com/adoptium/temurin-build/pull/3825#discussion_r1617336333

@karianna  - do you think that auto merge can be changed to auto rebase?

**Comment by sxa** [comment]
> Hmm, this is probably the Update Branch feature that was enabled on some repos and defaults to merge. @gdams do you have perms to see if you can make that rebase by default?

I really don't like that feature and I'm tempted to propose just switching it off entirely. It seems to have caused more problems that it has solved. In this case we've ended up with changes that seem to have been reverted in a single commit along with with the originator's changes so I can't see an easy way to unmerge them from here.

@suchismith1993 if you have a local copy of just the changes you've made, can you force push to this branch to get back to the changes you want? That might be the best course of action here.

File: build-farm/platform-specific-configurations/aix.sh [hunk 1]
```diff
@@ -133,7 +133,13 @@ if [ "$JAVA_FEATURE_VERSION" -ge 11 ]; then
   export CC=xlclang
   export CXX=xlclang++
 fi
-
+if [ "$JAVA_FEATURE_VERSION" -ge 22 ]; then
```
Comment (sxa) [inline build-farm/platform-specific-configurations/aix.sh:136]: Suggest changing this from `fi / if` to `elif`

File: build-farm/platform-specific-configurations/aix.sh [hunk 2]
```diff
@@ -127,7 +127,7 @@ else
   export CONFIGURE_ARGS_FOR_ANY_PLATFORM="${CONFIGURE_ARGS_FOR_ANY_PLATFORM} DF=/usr/sysv/bin/df"
 fi
 
-if [ "$JAVA_FEATURE_VERSION" -ge 11 ]; then
+if [ "$JAVA_FEATURE_VERSION" -le 21 ]; then
```
Comment (sxa) [inline build-farm/platform-specific-configurations/aix.sh:130]: We still build JDK8 with XLC13, so this should probably be something like:
   `if [ "$JAVA_FEATURE_VERSION" -le 21 -a "$JAVA_FEATURE_VERSION" -ge 11 ]; then`

**Comment by karianna** [comment]
@suchismith1993 unhappy linter :-)

**Comment by karianna** [comment]
Still a few to go:

In /github/workspace/build-farm/platform-specific-configurations/aix.sh line 130:
if [ [ "$JAVA_FEATURE_VERSION" -le 21 ] -a [ "$JAVA_FEATURE_VERSION" -ge 11 ] ]; then
^-- SC1009 (info): The mentioned syntax error was in this if expression.
   ^-- SC1073 (error): Couldn't parse this test expression. Fix to allow more checks.
     ^-- SC1026 (error): If grouping expressions inside [..], use \( ..\).
        ^-- SC1072 (error): Expected comparison operator (don't wrap commands in []/[[]]). Fix any mentioned problems and try again.

For more information:
  https://www.shellcheck.net/wiki/SC1026 -- If grouping expressions inside [....
  https://www.shellcheck.net/wiki/SC1072 -- Expected comparison operator (don...
  https://www.shellcheck.net/wiki/SC1073 -- Couldn't parse this test expressi...
------
2024-06-10 00:08:41 [INFO]   ---------------------------
2024-06-10 00:08:41 [INFO]   File:[/github/workspace/sbin/build.sh]
2024-06-10 00:08:42 [ERROR]   Found errors in [shellcheck] linter!
2024-06-10 00:08:42 [ERROR]   Error code: 1. Command output:
------

In /github/workspace/sbin/build.sh line 135:
      if [ [ "${BUILD_CONFIG[OS_KERNEL_NAME]}"  == "aix" ] && [ "${JDK_VERSION_NUMBER}" -lt 22 ] ]; then
      ^-- SC[100](https://github.com/adoptium/temurin-build/actions/runs/9440451983/job/25999781020?pr=3822#step:4:101)9 (info): The mentioned syntax error was in this if expression.
         ^-- SC1073 (error): Couldn't parse this test expression. Fix to allow more checks.
           ^-- SC[102](https://github.com/adoptium/temurin-build/actions/runs/9440451983/job/25999781020?pr=3822#step:4:103)6 (error): If grouping expressions inside [..], use \( ..\).
              ^-- SC1072 (error): Expected comparison operator (don't wrap commands in []/[[]]). Fix any mentioned problems and try again.

**Comment by suchismith1993** [comment]
@karianna  @sxa  Where can i check these linter errors ?

**Comment by karianna** [comment]
> @karianna @sxa Where can i check these linter errors ?

Check the result of the GitHub actions.  You can also run shellcheck locally.

**Comment by suchismith1993** [comment]
> > @karianna @sxa Where can i check these linter errors ?
> 
> Check the result of the GitHub actions. You can also run shellcheck locally.

Thanks. I hope it looks ok now ?

**Comment by sxa** [comment]
Testing:
- JDK8: https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk8u/job/jdk8u-aix-ppc64-temurin/466/
- JDK22: https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk22u/job/jdk22u-aix-ppc64-temurin/13
- JDK (latest): https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk/job/jdk-aix-ppc64-temurin/307/

File: sbin/build.sh [hunk 3]
```diff
@@ -132,7 +132,7 @@ configureReproducibleBuildParameter() {
 
       # Ensure reproducible and comparable binary with a unique build user identifier
       addConfigureArg "--with-build-user=" "admin"
-      if [ "${BUILD_CONFIG[OS_KERNEL_NAME]}" == "aix" ]; then
+      if [ "${BUILD_CONFIG[OS_KERNEL_NAME]}"  == "aix" ] && [ "${JDK_VERSION_NUMBER}" -lt 22 ]; then
```
Comment (sxa) [inline sbin/build.sh:135]: From the builds logs, it looks like JDK_VERSION_NUMBER isn't available here

Comment (suchismith1993) [reply sbin/build.sh:135]: How does the build pass for JDK8 then ?

Comment (suchismith1993) [reply sbin/build.sh]: I am facing this error too now, after removing extra brackets. I am using the OPENJDK map to access it now

**Comment by suchismith1993** [comment]
checking flags for boot jdk java command for small workloads...  -XX:+UseSerialGC -Xms32M -Xmx512M -XX:TieredStopAtLevel=1
checking for --enable-icecc... disabled, default
checking if precompiled headers are available... yes
checking for --enable-precompiled-headers... enabled, from default 'auto'
checking for ccache... [not found]
checking if ccache is available... no, ccache binary missing or not executable
checking if ccache is enabled... disabled, from command line
checking if build directory is on local disk... no
configure: creating /home/suchiroy/suchibuild/temurin-build/build-farm/workspace/build/src/build/aix-ppc64-server-release/configure-support/config.status
config.status: creating /home/suchiroy/suchibuild/temurin-build/build-farm/workspace/build/src/build/aix-ppc64-server-release/spec.gmk
config.status: creating /home/suchiroy/suchibuild/temurin-build/build-farm/workspace/build/src/build/aix-ppc64-server-release/bootcycle-spec.gmk
config.status: creating /home/suchiroy/suchibuild/temurin-build/build-farm/workspace/build/src/build/aix-ppc64-server-release/buildjdk-spec.gmk
config.status: creating /home/suchiroy/suchibuild/temurin-build/build-farm/workspace/build/src/build/aix-ppc64-server-release/compare.sh
config.status: creating /home/suchiroy/suchibuild/temurin-build/build-farm/workspace/build/src/build/aix-ppc64-server-release/Makefile

====================================================
A new configuration has been successfully created in
/home/suchiroy/suchibuild/temurin-build/build-farm/workspace/build/src/build/aix-ppc64-server-release
using configure arguments '--verbose --with-vendor-name='Undefined Vendor' --with-vendor-url=file:///dev/null --with-vendor-bug-url=file:///dev/null --with-vendor-vm-bug-url=file:///dev/null --with-version-opt=202406110845 --with-version-pre=beta --with-version-build=8 --with-boot-jdk=/home/suchiroy/suchibuild/temurin-build/build-farm/jdk-21 --with-debug-level=release --with-native-debug-symbols=external --with-alsa=/home/suchiroy/suchibuild/temurin-build/build-farm/workspace/./build//installedalsa --with-source-date=1718095513 --with-hotspot-build-time='2024-06-11 08:45:13' --disable-ccache --with-build-user=admin --with-jvm-variants=server --with-cacerts-src=/home/suchiroy/suchibuild/temurin-build/sbin/../security/certs --disable-warnings-as-errors --with-cups-include=/opt/freeware/include DF=/usr/sysv/bin/df --with-memory-size=7000 --with-freetype=bundled --with-zlib=bundled'.

Configuration summary:
* Name:           aix-ppc64-server-release
* Debug level:    release
* HS debug level: product
* JVM variants:   server
* JVM features:   server: 'compiler1 compiler2 epsilongc g1gc jfr jni-check jvmti management parallelgc serialgc services vm-structs' 
* OpenJDK target: OS: aix, CPU architecture: ppc, address length: 64
* Version string: 22.0.1-beta+8-202406110845 (22.0.1-beta)
* Source date:    1718095513 (2024-06-11T08:45:13Z)

Tools summary:
* Boot JDK:       openjdk version "21.0.3" 2024-04-16 LTS OpenJDK Runtime Environment Temurin-21.0.3+9 (build 21.0.3+9-LTS) OpenJDK 64-Bit Server VM Temurin-21.0.3+9 (build 21.0.3+9-LTS, mixed mode) (at /home/suchiroy/suchibuild/temurin-build/build-farm/jdk-21)
* Toolchain:      clang (IBM Open XL C/C++)
* C Compiler:     Version 15.0.0 (at /opt/IBM/openxlC/17.1.1/bin/ibm-clang_r)
* C++ Compiler:   Version 15.0.0 (at /opt/IBM/openxlC/17.1.1/bin/ibm-clang++_r)

Build performance summary:
* Build jobs:     6
* Memory limit:   7000 MB

WARNING: Your build output directory is not on a local disk.
This will severely degrade build performance!
It is recommended that you create an output directory on a local disk,
and run the configure script again from that directory.

The following warnings were produced. Repeated here for convenience:
WARNING: alsa not used, so --with-alsa[-*] is ignored

Building targets 'product-images test-image static-libs-graal-image' in configuration 'aix-ppc64-server-release'
Compiling up to 8 files for BUILD_TOOLS_LANGTOOLS
Compiling up to 1 files for BUILD_TOOLS_HOTSPOT
Creating hotspot/variant-server/tools/adlc/adlc from 1

**Comment by suchismith1993** [comment]
Build proceeds to compilation now.

**Comment by sxa** [comment]
You seem to have lost the `#` on the `#!/bin/bash` at the start of the file - can you reinstate it please?

**Comment by sxa** [comment]
Have we had any confirmation about the behaviour of `-qnotimestamps` with OpenXL17 i.e. whether it is the default now, or whether there is another way to enable comparable functionality?

**Comment by suchismith1993** [comment]
> Have we had any confirmation about the behaviour of `-qnotimestamps` with OpenXL17 i.e. whether it is the default now, or whether there is another way to enable comparable functionality?
I did not find a suitable alternate option from the help page.

**Comment by sxa** [comment]
Re-running build pipelines:
- JDK8: https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk8u/job/jdk8u-aix-ppc64-temurin/467/
- JDK22: https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk22u/job/jdk22u-aix-ppc64-temurin/14/ (Note: will NOT run the tests on the build machine, so they will likely fail with a missing OpenXL17 runtime. This build was also done after rebasing on the upstream temurin-build master branch)
- JDK head: https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk/job/jdk-aix-ppc64-temurin/316/ (Currently queued up, also set to run the tests on the build machine for now so it has the OpenXL17 runtime available)

**Comment by sxa** [comment]
The JDK head pipeline is showing [this failure](https://github.com/adoptium/aqa-tests/issues/4429) in [the foreign TestLargeSegmentCopy test](https://ci.adoptium.net/job/Test_openjdk24_hs_sanity.openjdk_ppc64_aix_rerun/1/testReport/java_foreign_TestLargeSegmentCopy/java/TestLargeSegmentCopy/)

**Comment by sxa** [comment]
Sanity.openjdk run against the JDK22 build pipeline from the [earlier comment](https://github.com/adoptium/temurin-build/pull/3822#issuecomment-2160247401): https://ci.adoptium.net/job/Grinder/10361/console

**Comment by sxa** [comment]
> Sanity.openjdk run against the JDK22 build pipeline from the [earlier comment](https://github.com/adoptium/temurin-build/pull/3822#issuecomment-2160247401): https://ci.adoptium.net/job/Grinder/10361/console

PASSED

Running extended.openjdk at https://ci.adoptium.net/job/Grinder/10371/

**Comment by sxa** [comment]
Noting that we'll also need a PR into https://github.com/adoptium/ci-jenkins-pipelines/blob/49d73e26f7ff74a7872d00b856af003745ef2dbb/pipelines/jobs/configurations/jdk22u_pipeline_config.groovy#L77 (and later versions) to make these builds use a build machines labelled as `xlc17` label instead of `xlc16` (or maybe `openxl17` would be more appropriate? We can use whatever makes most sense from IBM's perspective) which will tie it to the new machine. Similarly we should probably use `sw.os.aix.7_2&&openxl17` in the `additionalTestLabels` for now.
FYI @andrew-m-leonard

**Comment by sxa** [comment]
> Running extended.openjdk at https://ci.adoptium.net/job/Grinder/10371/

Job timed out after 10 hours - it would be good to check whether anything failed in that job during those ten hours, but I've kicked off abother run with a 100 hour timeout at https://ci.adoptium.net/job/Grinder/10377/

**Comment by sxa** [comment]
> EDIT: 10377 failed due to the error referenced in [this issue](https://github.com/adoptium/aqa-tests/pull/5374#issuecomment-2164303424) and fixed by a revert in [adoptium/aqa-tests#5386](https://github.com/adoptium/aqa-tests/pull/5386) so I've kicked off a simple re-run of 10377 at https://ci.adoptium.net/job/Grinder/10383/

Completed without failures in 18 hours 👍🏻

**Review by sxa — APPROVED** [review]
I think is is likely good to go now, although as mentioned before it will require a change in ci-jenkins-pipelines to ensure the builds and tests are sent to the correct machine.

Looking at the time stamps, the option was originally added as part of https://github.com/adoptium/temurin-build/issues/3375#issuecomment-1568490734 - looking at two object files for a "hello world" app created using OpenXL17 it looks like there are no differences other than the timestamps near the start of the file so it is likely that the option is now default behaviour.

The following is the comparison of a file created from `od -cx` on the files to easily show the differences:
```
sxa@adopt09:[/home/sxa]diff hello?.o.cx
1,2c1,2
< 0000000     01df    0002    666c    5228    0000    0102    0000    0013
< 	 001 337  \0 002   f   l   R   (  \0  \0 001 002  \0  \0  \0 023
---
> 0000000     01df    0002    666c    522c    0000    0102    0000    0013
> 	 001 337  \0 002   f   l   R   ,  \0  \0 001 002  \0  \0  \0 023
sxa@adopt09:[/home/sxa]
```
Same for the final built binary:
```
sxa@adopt09:[/home/sxa]diff hello?.o.cx
1,2c1,2
< 0000000     01df    0002    666c    5228    0000    0102    0000    0013
< 	 001 337  \0 002   f   l   R   (  \0  \0 001 002  \0  \0  \0 023
---
> 0000000     01df    0002    666c    522c    0000    0102    0000    0013
> 	 001 337  \0 002   f   l   R   ,  \0  \0 001 002  \0  \0  \0 023
sxa@adopt09:[/home/sxa]
```

**Comment by karianna** [comment]
@suchismith1993 Let us konw when it's safe to merge alongside the CI changes

**Comment by suchismith1993** [comment]
> @suchismith1993 Let us konw when it's safe to merge alongside the CI changes

I am not sure about the CI changes . I think that is a separate repository ? I think @sxa  or @Haroon-Khel  can comment better on that.

**Comment by sxa** [comment]
> > @suchismith1993 Let us konw when it's safe to merge alongside the CI changes
> 
> I am not sure about the CI changes . I think that is a separate repository ? I think @sxa or @Haroon-Khel can comment better on that.

Those will be the ones I mentioned in my [earlier comment](https://github.com/adoptium/temurin-build/pull/3822#issuecomment-2163436165)

**Comment by sxa** [comment]
Ref [earlier comment](https://github.com/adoptium/temurin-build/pull/3822#issuecomment-2163436165) I've added the `openxl17` and `sw.os.aix.7_2TL5` labels to the build machine for now, so those can be the basis of a PR to ci-jenkins-pipelines. @smlambert Based on our discussion earlier today let me know if you're happy with that label (The current ones are the same without `TL5` on the end)

**Pull request merged**